ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -208,12 +208,8 @@ private PullQueryRow pollRow() {
 
     polled.handler.handle(new SucceededFuture<>(null, null));
 
-    if (monitor.enterIf(atHalfCapacity)) {
-      try {
-        drainHandler.forEach(h -> h.handle(null));
-      } finally {
-        monitor.leave();
-      }
+    if (isDone() || size() <= queueCapacity / 2) {
+      drainHandler.forEach(h -> h.handle(null));
     }
```

Comment:
I think the drain handler is called too often also in the original code. As far as I understand the drain handler should be called to resume the source that is piped into the write stream (i.e. the http response) after is has been paused. Just because the size of the queue is less than half its capacity, it does not mean the http response has been paused, it could also be that the response gets content slower.
During our investigation, I tried to call the drain handler only when the http response has been paused before and that has already reduced the error rate significantly but did not solve the issue, though.
What do you think of also reducing the calls to the drain handler?
Probably, we should not do it in this PR since for testing your fix we want to hit the issue often. Maybe let's consider it an optimization that we might or might not add afterwards.

Reply:
Yes, I think the whole way the drain handler is used is quite mysterious to me. It's called too often, it's protected by a monitor whose purpose is not clear, it's called from a different thread (other WriteStreams call it from `write`, so basically from the same context).

I think you are right that it should be called less often, however this change is really to do the minimum change necessary. It's an idea to try this in a separate PR, but then again, not sure if there is enough interest to improve pull queries at this point.